docs/language/crypto.mdx

```diff
@@ -0,0 +1,333 @@
+   A signature is 48-bytes long.
+   - `signedData` is the arbitrary message to verify the signature against.
+   - `domainSeparationTag` is the domain tag used for signing. All tags are accepted. 
+   - `hashAlgorithm` is the algorithm used to hash the message along with the given tag (check `hashWithTag` function for more details). 
+   Only `KMAC128_BLS_BLS12_381` is accepted.
+
+BLS verification performs the necessary membership check of the signature while the membership check of the public key is performed at the creation of the `PublicKey` object. 
+
+## Crypto Contract
+
+The built-in contract `Crypto` can be used to perform cryptographic operations.
+The contract can be imported using `import Crypto`.
+
+### Hashing
+
+The crypto contract hashing API is:
+
+```cadence
+// Hash the data using the given hashing algorithm and returns the hashed data.
+pub fun hash(_ data: [UInt8], algorithm: HashAlgorithm): [UInt8]
+
+// Hash the data using the given hashing algorithm and the tag. Returns the hashed data.
+pub fun hashWithTag(_ data: [UInt8], tag: string, algorithm: HashAlgorithm): [UInt8]
+```
+
+<Callout type="info">
+
+🚧 Status: `KMAC128_BLS_BLS12_381` is not supported yet.
+
+</Callout>
+
+- `hash` hashes the input data using the input hashing algorithm. 
+    - If `KMAC128_BLS_BLS12_381` is used, data is hashed using the standard mac `KMAC128(customizer, key, data, length)` with the following inputs:
+        - `customizer` is the UTF-8 encoding of `"H2C"` 
+        - `key` is the UTF-8 encoding of `"APP_BLS_SIG_BLS12381G1_XOF:KMAC128_SSWU_RO_POP_"`
+        - `data` is the input data to hash
+        - `length` is 1024 bytes
+
+<Callout type="info">
+
+🚧 Status: `hashWithTag` is not supported yet.
+
+</Callout>
+
+- `hashWithTag` hashes data along with a tag. This allows instanciating independent hashing functions customized with a domain separation tag. This is implemented differently depending on the hashing algorithm:
+    - `SHA2_256`, `SHA2_384`, `SHA3_256`, `SHA3_384` : the hashed message is `bytes(tag) || data` where `bytes()` is the UTF-8 encoding of the input string, padded with zeros till 32 bytes. The tags accepted must not exceed 32 bytes.
+    - `KMAC128_BLS_BLS12_381` : this is a call to standard `KMAC128(customizer, key, data, length)` with the following inputs:
+        - `customizer` is the UTF-8 encoding of `"H2C"`
+        - `key` is the UTF-8 encoding of `"APP_" || tag || "BLS_SIG_BLS12381G1_XOF:KMAC128_SSWU_RO_POP_"`
+        - `data` is the input data to hash
+        - `length` is 1024 bytes
+
+
+For example, to compute a SHA3-256 hash:
+
+```cadence
+let data: [UInt8] = [1, 2, 3]
+let digest = Crypto.hash(data, algorithm: HashAlgorithm.SHA3_256)
+```
+
+The crypto contract also allows creating key lists to be used for a simple multi-signature 
+verification. 
+For example, to verify two signatures with equal weights for some signed data:
```

Comment:
Just realized that this code sample showing how to use key lists to do multi-sig verification is now under the Hashing subsection, add a subsection here:

```suggestion

### Key Lists

The crypto contract also allows creating key lists to be used for multi-signature verification.
For example, to verify two signatures with equal weights for some signed data:
```